RHOAIENG-3458: Introduce edge client and model registry client

devguyio opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
## Description
<!--- Describe your changes in detail -->
Using the model registry to implement the edge use case includes a lot of custom logic. 

This PR adds a model registry client and an edge client as two abstractions that encapsulate such logic and can be used by the CLI directly. 

This PR also refactors the models command to use the newly introduced edge client and streamlines the command design.
>[!TIP]
> This PR is reviewed best commit-by-commit. Instead of using the `Files changed` tab:
> - Use the `Commits` tab
> - Click on the top commit and review that commit
> - Use `n` and `p` to go to the next and previous commits respectively to review each commit. 


## How Has This Been Tested?
```shell
cd cli
make cli-build
./odh models
```
![image](https://github.com/opendatahub-io/ai-edge/assets/1930204/8c0199b7-fea1-4516-885e-9dd594de4f84)

```shell
./odh models add "voice-recognition-model" "A ML model for voice recognition" "v1"   
```
![image](https://github.com/opendatahub-io/ai-edge/assets/1930204/8a0b25c2-020b-4cfb-a6af-53847973729d)


## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [x] The commits are squashed in a cohesive manner and have meaningful messages.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work

devguyio commented:
@biswassri FYI

Sara4994 commented:
Can we have an Readme containing instructions to build and run the cli, also list of available commands and its syntax? it will easy for others to try it out. wdyt?

grdryn commented:
/lint

grdryn commented on cli/params.yaml:
```diff
@@ -0,0 +1,34 @@
+params:
+  - name: s3-bucket-name
+    value: rhoai-abdalla-pipelines
```
This seems quite specific to you. Is this file even expected to be committed here?

devguyio replied:
Indeed it wasn't supposed to be committed. I changed it to be a sample file that's under `cli/examples` and added instructions in the `docs/cli.md` documentation.

grdryn commented on cli/params.yaml:
```diff
@@ -0,0 +1,34 @@
+params:
```
Is my understanding correct that all of the data here would be added as metadata to a model in the model registry? I wonder if the pipeline should be able to figure out some of the information based on standard model information in the registry, rather than having metadata for each pipeline parameter. :thinking: 

Having not thought about it too much, I wonder if the _output_ of the MLOps Pipeline (image reference) is more relevant information to store with a model, rather than the pipeline _input_ params?

devguyio replied:
> Is my understanding correct that all of the data here would be added as metadata to a model in the model registry? I wonder if the pipeline should be able to figure out some of the information based on standard model information in the registry, rather than having metadata for each pipeline parameter. 🤔

- The implemented CLI use case stores the metadata once and then reuses it whenever an image build is triggered. 
- I believe some metadata can be figured out by the pipeline, and some need to be stored somewhere. E.g. the location of the model.
- I see it as inevitable that such metadata shouldn't be stored in the current naive format of arbitrary key-value pairs. Instead, the MR should have some properties that have a semantic meaning that the pipeline and/or CLI can use. However, the current approach seems okay for proofing the concept/demo purposes.

> 
> Having not thought about it too much, I wonder if the _output_ of the MLOps Pipeline (image reference) is more relevant information to store with a model, rather than the pipeline _input_ params?

Well it depends on where we're gonna end up storing the model container image information, cause one option that's up for future investigation is to use the container registry and [OCI 1.1 to manage the artifacts metadata](https://github.com/opencontainers/image-spec/blob/main/manifest.md#guidelines-for-artifact-usage)  and in such case I can imagine a lot of the required metadata can be available in the container registry or have to be updated by the pipeline at some stage.

Currently, it'd definitely make sense to make the pipeline update the status of the model artifact and version in the model registry with the pipeline build result and the container image sha etc.

grdryn commented on cli/params.yaml:
```diff
@@ -0,0 +1,34 @@
+  - name: git-containerfile-revision
+    value: "main"
+  - name: containerfileRelativePath
+    value: pipelines/containerfiles/Containerfile.seldonio.mlserver.mlflow
```
As one example, I'm not sure if this would fit as _model_ metadata, since a hypothetical user may want to package the same model into two different images with different model servers (think of deployment to two different edge locations, with slightly different hardware, or something), which would mean two different images would be created from two different Containerfiles.

I don't have a great handle on the hierarchy of things in the model registry yet though, and where this fits in; so what I'm saying might not make 100% sense.

devguyio replied:
> As one example, I'm not sure if this would fit as _model_ metadata, 

As I mentioned above, I think we should aim to have a properly designed API that contains metadata with proper semantic meanings. So, this definitely needs to go back to the whiteboard, as currently, it's just a bunch of arbitrary key-value pairs.

> since a hypothetical user may want to package the same model into two different images with different model servers (think of deployment to two different edge locations, with slightly different hardware, or something), which would mean two different images would be created from two different Containerfiles.

I see your point. However, if we're talking about container images that combine a model server and the model, then I believe each variant should be considered a different artifact for the model. In such cases, some of the metadata will be associated only with a specific kind of artifact, and others will stay at the version level.

grdryn reviewed (commented):
Is the JIRA ID correct in the title? It appears to reference a closed issue RHOAIENG-3458

grdryn commented on cli/pkg/edgeclient/types.go:
```diff
@@ -0,0 +1,42 @@
+
+package edgeclient
+
+type ModelImageStatus int
+
+func (s ModelImageStatus) String() string {
+	return [...]string{"Unknown", "Needs Sync", "Synced", "Building", "Live", "Failed"}[s]
+}
```
For my own understanding, do you have any links to docs on this kind of construct, or a name of a pattern or something that I could search for? I don't remember having ever encountered it before (It has been more years since I last used Go than I actually used it for, so I might just have forgotten).

devguyio replied:
> For my own understanding, do you have any links to docs on this kind of construct, or a name of a pattern or something that I could search for? I don't remember having ever encountered it before (It has been more years since I last used Go than I actually used it for, so I might just have forgotten).

@grdryn So, this was supposed to be an enum, Go's version of it anyway, but the `ModelImageStatus` was not the correct reference to that since when I split it, I forgot to introduce the constants. 

Here's a good article about [implementing Enum's in Go](https://dizzy.zone/2024/01/26/Enums-in-Go/).

grdryn replied:
I had a feeling it was something around enums, and Go's lack of useful features. 👍

grdryn commented:
> This PR is reviewed best commit-by-commit. Instead of using the Files changed tab [...]

One drawback of this approach is that I don't seem to be able to leave inline suggestions as easily on all files; and also I don't seem to be able to mark all files as viewed, although I can for some (I presume if a file is changed in more than one commit in the PR, it can't be marked as viewed in any single one of those).

grdryn commented:
/remove-approve

(new process)

devguyio commented:
/lint

devguyio commented:
/lint

devguyio commented on cli/pkg/commands/common/types.go:
```diff
@@ -0,0 +1,36 @@
+var MessageStyle = lipgloss.NewStyle().
+	Bold(true)
+
+var ErrorStyle = lipgloss.NewStyle().
```
Done

devguyio commented on cli/pkg/commands/common/cmd.go:
```diff
@@ -0,0 +1,76 @@
+	"github.com/spf13/cobra"
+)
+
+type SubCommand int
```
Check the new structure

devguyio commented on cli/pkg/edgeclient/client.go:
```diff
@@ -0,0 +1,108 @@
+// Client is a client representing the edge environment
+//
+// This client can be used to create and manage models and model container images suitable for deployment in edge environments.
+type Client struct {
```
The main types used in a go file are recommended to stay in the same file next to the method receivers. So while we can have a `type.go` file that includes multiple, small types. The `Client` struct is the main type in this package and it should stay in that file next to the constructor and method receivers.

devguyio commented:
> Can we have an Readme containing instructions to build and run the cli, also list of available commands and its syntax? it will be easy for others to try it out. wdyt?

Added a readme.

devguyio commented on cli/pkg/commands/models/add.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+func init() {
+	addCmd.Flags().StringP("params", "p", "params.yaml", "Path to the build parameters file")
```
Done

devguyio commented on cli/pkg/pipelines/params.go:
```diff
@@ -0,0 +1,70 @@
+	"gopkg.in/yaml.v2"
+)
+
+type Params struct {
```
Definitely!!

devguyio commented on cli/pkg/modelregistry/client.go:
```diff
@@ -0,0 +1,298 @@
+// Client is a client for the model registry service
+type Client struct {
+	// modelRegistryURL is the URL of the model registry service
+	modelRegistryURL    string
```
IIRC I put it there for when needing to debug the client. Yeah, I'd prefer that we keep it. It's an important detail that makes sense to be available in the client IMHO.

devguyio commented on cli/pkg/modelregistry/client.go:
```diff
@@ -0,0 +1,298 @@
+		return nil, nil, nil, fmt.Errorf("name, description and version are required")
+	}
+
+	m, err := c.CreateRegisteredModel(modelName, modelDescription, nil)
```
Correct

devguyio commented on cli/pkg/modelregistry/errors.go:
```diff
@@ -0,0 +1,77 @@
+	return "GenericOpenAPIError"
+}
+
+func (e GenericOpenAPIError) Is(target error) bool {
```
Correct, I forgot to remove it from the PR. Removed!

devguyio commented:
> Is the JIRA ID correct in the title? It appears to reference a closed issue RHOAIENG-3458

@grdryn The JIRA ID is correct, this can be considered a follow-up work relevant to that JIRA, which got closed when #227 was merged. Both PRs are about the models command.

devguyio commented on cli/pkg/modelregistry/errors.go:
```diff
@@ -0,0 +1,77 @@
+	return false
+}
+
+type GenericOpenAPIError struct {
```
This one was a left over by mistake. I removed it.

devguyio commented on cli/pkg/modelregistry/errors.go:
```diff
@@ -0,0 +1,77 @@
+// isOpenAPIErrorOfKind checks if the error is of the given kind (targetErr). It checks if sourceErr is an
+// openapi.GenericOpenAPIError and if the error message contains the targetErr message.
+//
+// This is a workaround to handle the error until the model registry supports returning standard HTTP status codes for
+// errors with known status codes.
```
It is planned indeed and I created #244 to track that

devguyio commented on cli/pkg/modelregistry/client.go:
```diff
@@ -0,0 +1,298 @@
+			return nil, fmt.Errorf("error creating model version: %w", err)
+		} else {
+			if resp.StatusCode != 201 {
+				// TODO: Remove this workaround when model registry returns 404 when the model is not found
```
Do you mean linked in the source code?

grdryn replied:
Yeah, possibly. I think linking issues in the source code is useful, so that when we stumble across this TODO in future, we'll have concrete places to look to see if it can be progressed.

devguyio commented on cli/pkg/modelregistry/client.go:
```diff
@@ -0,0 +1,298 @@
+	a, resp, err := c.modelRegistryClient.ModelRegistryServiceAPI.CreateModelVersionArtifact(
+		context.Background(), versionId,
+	).
+		Artifact(openapi.ModelArtifactAsArtifact(artifact)).Execute()
```
Yeah, definitely, that was an oversight. Fixed.

devguyio commented on cli/pkg/edgeclient/client.go:
```diff
@@ -0,0 +1,108 @@
+}
+
+// AddNewModelWithImage adds a model to the model registry along with the model version and all build parameters required to
+// create a model container image using a tekton pipeline run.
```
No it doesn't. I fixed the comment.

devguyio commented on cli/pkg/modelregistry/client.go:
```diff
@@ -0,0 +1,298 @@
+) (*openapi.ModelVersion, error) {
+
+	if modelId == "" || versionName == "" {
+		return nil, fmt.Errorf("model ID and version are required")
```
Done

devguyio commented:
> We should also include this in the Makefile and in the PR check as well under the `make test` target.

@MarianMacik can you please clarify what should be included in the Makefile?

Sara4994 commented:
@devguyio There is a confusion around JIRA number for this PR. I see [RHOAIENG-3458](https://issues.redhat.com/browse/RHOAIENG-3458) is closed already.

LaVLaS commented on docs/cli.md:
```diff
@@ -0,0 +1,135 @@
+- Go (version 1.20 or 1.21, version 1.22 is not supported).
+- [Model registry operator](https://github.com/opendatahub-io/model-registry-operator/releases/tag/v0.1.2) version 0.1.2 locally.
+
+### Installing Model Registry
```
With ODH release >= v2.9, can these initial instructions (Steps 1-5) for installing the model-registry be simplified to installing ODH with the Model Registry [operator] component enabled and deploy from there?

devguyio replied:
@LaVLaS I haven't tested it against that version as at the time this PR was open, that was not there. Besides, this might end up with the need to bump up the REST client and test the whole CLI commands against it.

This was developed against Model Registry `0.1.2` which was the working version. How do you feel about merging this PR as is,  since `0.1.2`  is a release of Model Registry, and I can create a follow-up issue to bump up the model registry version and update the instructions?

grdryn commented on docs/cli.md:
````diff
@@ -0,0 +1,135 @@
+    ```bash
+    cd model-registry-0.1
+    ```
+2. Install the CRDs into the cluster:
````
Should we consider just pointing to the model registry operator repo for the instructions to install that, rather than having our own?

grdryn commented on docs/cli.md:
```diff
@@ -0,0 +1,135 @@
+
+- `-m, --model-registry-url`: The URL of the Model Registry. This is the URL of the Model Registry service that is
+  running in the OpenShift cluster.
+- `k, --kbueconfig` : The path to the kubeconfig file. This is the kubeconfig file that is used to access the OpenShift 
```
```suggestion
- `k, --kubeconfig` : The path to the kubeconfig file. This is the kubeconfig file that is used to access the OpenShift 
```

devguyio commented on docs/cli.md:
```diff
@@ -0,0 +1,135 @@
+### Prerequisites
+
+- Kubectl access to the OpenShift core cluster.
+- Go (version 1.20 or 1.21, version 1.22 is not supported).
```
Model Registry builds and uses `controller-gen` to create some manifests. The  version used crashes with go 1.22 golang/go#65637 . Once we bump up the model registry version (which is not a straight forward task) we'll probably get the newest version of `controller-gen` and will be okay.

devguyio commented on docs/cli.md:
````diff
@@ -0,0 +1,135 @@
+1. Change to the `model-registry-operator-0.1.2` directory:
+
+    ```bash
+    cd model-registry-0.1
````
Fixed

devguyio commented on cli/pkg/commands/models/models.go:
```diff
@@ -0,0 +1,176 @@
+/*
+Copyright 2023 KStreamer Authors
```
Fixed

devguyio commented on cli/pkg/modelregistry/client.go:
```diff
@@ -0,0 +1,298 @@
+	a, err := c.CreateModelArtifact(
+		v.GetId(), artifactName, modelDescription, uri, modelFormatName, modelFormatVersion, artifactExternalId,
+	)
+
```
Fixed

Sara4994 commented:
/lgtm

Sara4994 commented:
This PR looks good to me, leaving this to Gerard and Landon now :)

devguyio commented:
Holding for other reviewers 

/hold

devguyio commented:
> This PR looks good to me, leaving this to Gerard and Landon now :)

And @MarianMacik

grdryn reviewed (approved):
/lgtm

LaVLaS commented:
/unhold

MarianMacik commented on docs/cli.md:
```diff
@@ -0,0 +1,135 @@
+
+### Adding a Model
+
+Adding a model to the model registry means adding the metadata about the model to the Model Registry. This metadata includes the model name, a model version and the pipeline parameters that are required to build the model image.
```
Should we also mention the third part, which is the model artifact? By default the CLI will create an artifact with the same name as the model.

devguyio replied:
I intentionally skipped documenting how we're using the model registry in this document and focused on the usage of the CLI from the user perspective. For the user, he provides a model name, the version, and the pipeline run parameters regardless of how they're stored in the model registry. The artifact concept isn't exposed to the CLI users.

MarianMacik reviewed (commented):
I have just a few more comments. Nothing blocking, though. Thanks for all the changes, the code looks even better now!

devguyio commented on cli/pkg/commands/common/cmd.go:
```diff
@@ -0,0 +1,106 @@
+		},
+	}
+
+	for _, f := range flags {
+		if !f.IsRootFlag() {
+			if f.IsInherited() {
+				cmd.PersistentFlags().StringP(f.String(), f.Shorthand(), f.Value(), f.Usage())
+			} else {
+				cmd.Flags().StringP(f.String(), f.Shorthand(), f.Value(), f.Usage())
+			}
+		}
+	}
```
Here we're registering non-root flags only. Root flags are registered in `pkg/commands/root.go`. The `FlagsModelRegistryURL` is a root flag as declared in `pkg/commands/flags/flags.go`. 

One logical error that I just fixed, is that root flags should also be set as inherited. I pushed a fix. It wasn't producing a bug but it was logically incorrect. 

A root flag means it's defined at the root level, so can be inherited by all subcommand.

There can be a case of an inherited flag that's not root. In that case, it's defined by a specific sub-command and will be inherited by all of its nested sub-commands.

MarianMacik replied:
@devguyio I see, I didn't realize that it is a root flag :)

However, in case of an inherited flag that is not root, if we add the third layer of commands, such a flag will be re-added at the third layer to the persistent flags, right? Because it will need to be listed among the flags to read and because it is the same slice, it will be iterated over when registering as well.
Right now it is not an issue as everything non-root should be registered. Also, maybe it won't be an issue even afterwards because cobra maybe can handle it.

devguyio commented on cli/examples/params.yaml.sample:
```diff
@@ -0,0 +1,52 @@
+    value: CONFIG_MAP_NAME
+    # Whether to add a workspace directory to the pipeline. Can be set to "true" or "false"
+  - name: addDirWorkspace
+    value: "true"
```
Done.

MarianMacik commented on cli/pkg/commands/models/models.go:
```diff
@@ -0,0 +1,176 @@
+
+This command will list all the registered models available in the Open Data Hub model registry.`,
+	cobra.NoArgs,
+	[]flags.Flag{flags.FlagModelRegistryURL.SetInherited()},
```
```suggestion
	[]flags.Flag{flags.FlagModelRegistryURL},
```

devguyio replied:
Fixed

MarianMacik commented:
> > We should also include this in the Makefile and in the PR check as well under the `make test` target.
> 
> @MarianMacik can you please clarify what should be included in the Makefile?

@devguyio I meant to include cli tests under the root makefile test target.

devguyio commented:
> > > We should also include this in the Makefile and in the PR check as well under the `make test` target.
> > 
> > 
> > @MarianMacik can you please clarify what should be included in the Makefile?
> 
> @devguyio I meant to include cli tests under the root makefile test target.

@MarianMacik I've pushed the following changes:
 - Added the cli-test as part of the root test target
 - Changed the flags packaged by renaming the `rootFlag` to `parentFlag` and now it means that this is a flag that will be declared by a parent so we're only interested in getting its value, not in declaring it.
 - Fixed a bug that was preventing adding two version with two different names under the same model:
   - `pkg/modelregistry` package
     -  Added `FindRegisteredModel` & `FindModelVersion`
     - Updated  `AutoRegisterModelVersionArtifact` method to check if the model or version exist first, and create them only if they don't exist.

MarianMacik reviewed (approved):
Nice catch with that bug, thanks for incorporating my comments!
/lgtm

devguyio commented:
/override ci/prow/test-ai-edge

grdryn reviewed (approved):
/approve
